lib/templateFactory.js

```diff
@@ -80,32 +81,60 @@ class TemplateFactory extends BaseFactory {
     }
 
     /**
-     * Get a the latest template by config
+     * Get a the latest template by config using the full template name
      * @method getTemplate
-     * @param  {Object}     config               Config object
-     * @param  {String}     config.name          The template name
-     * @param  {String}     config.version       Version of the template
-     * @param  {String}     [config.label]       Label of template
-     * @return {Promise}                         Resolves template model or undefined if not found
+     * @param  {String}     fullTemplateName    Name of the template and the version or tag (e.g. chef/publish@1.2.3)
+     * @return {Promise}                        Resolves template model or null if not found
      */
-    getTemplate(config) {
-        // get all templates with the same name
-        return super.list({ params: { name: config.name } })
+    getTemplate(fullTemplateName) {
+        const match = TEMPLATE_NAME_REGEX.exec(fullTemplateName);
+        const templateName = match[TEMPLATE_NAME_MATCH];
+        const versionOrTag = match[VERSION_OR_TAG_MATCH];
+        const isVersion = VERSION_REGEX.exec(versionOrTag);
+
+        // If tag is passed in, get the version from the tag
+        if (versionOrTag && !isVersion) {
+            // Lazy load factory dependency to prevent circular dependency issues
+            // eslint-disable-next-line global-require
+            const TemplateTagFactory = require('./templateTagFactory');
```

Comment:
I don't think putting lazy-loads in the constructor actually solves the circular dependency problem. If we have modules `A` and `B` that depend on each other we'd end up with constructor `A` calling require for `B`, which calls constructor `B`, which calls require for `A`, and so on.